// kem/lib/cookie/server.js
var querystring = require('querystring');

module.exports = class Cookie {
	constructor(string, res) {
		this.define = '';
		this.data = querystring.parse(string, '; ') || {};
		this.res = res;
	}

	encode(i) {
		return encodeURIComponent(i);
	}
	decode(i) {
		return decodeURIComponent(i);
	}

	init(set) {
		var date = new Date(),
		    time = set.day || 1,
		    path = set.path || '/';
		date.setTime(date.getTime() + time * 24 * 3600000);

		var array = ['expires=' + date.toUTCString(), 'path=' + path];
		if (set.host) array.push('domain=' + set.host);
		if (set.secure) array.push('secure=' + set.secure);
		if (set.http) array.push('httpOnly=' + set.http);

		this.define = array.join(';');
		return this;
	}

	set(array) {
		var list = [];

		for (var i in array) {
			var key = this.decode(i), value = this.decode(array[i]);
			list.push(key + '=' + value + ';' + this.define);
			this.data[key] = value;
		};

		this.res.setHeader('Set-Cookie', list);
	}

	has(name) {
		return this.data[name] !== '' && this.data[name] !== null;
	}

	get(name) {
		return this.data[name];
	}

	remove(...array) {
		if (!this.define) this.init({ day: -365, path: '/' });

		var list = [];

		for (var i = 0, n = array.length; i < n; i++) {
			list.push(this.decode(array[i]) + '=;' + this.define);
			delete this.data[this.decode(array[i])];
		};

		this.res.setHeader('Set-Cookie', list);
	}

	removeAll() {
		var list = [];

		for (var i in this.data) {
			list.push(i + '=;' + this.define);
		};

		this.res.setHeader('Set-Cookie', list);

		this.data = {};
	}
}
